access/common/auth/token_provider.go

```diff
@@ -0,0 +1,221 @@
+		c.Clock = clockwork.NewRealClock()
+	}
+	if c.Log == nil {
+		c.Log = logrus.NewEntry(logrus.StandardLogger())
+	}
+	return nil
+}
+
+// RotatedAccessTokenProvider is an implementation of AccessTokenProvider
+// that uses OAuth2 refresh token flow to renew the acess token.
+// The credentials are stored in the given persistent store.
+//
+// To have an up-to-date token, one must run RefreshLoop() in a background goroutine.
+type RotatedAccessTokenProvider struct {
+	ctx                 context.Context
+	retryInterval       time.Duration
+	tokenBufferInterval time.Duration
+	state               state.State
+	refresher           oauth.Refresher
+	clock               clockwork.Clock
+
+	log logrus.FieldLogger
+
+	lock  sync.RWMutex // protects the below fields
+	creds *state.Credentials
+}
+
+// NewRotatedTokenProvider creates a new RotatedAccessTokenProvider from the given config.
+// NewRotatedTokenProvider will return an error if the state does not have existing credentials,
+// meaning they need to be acquired first (e.g. via OAuth2 authorization code flow).
+func NewRotatedTokenProvider(cfg RotatedAccessTokenProviderConfig) (*RotatedAccessTokenProvider, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	provider := &RotatedAccessTokenProvider{
+		ctx:                 cfg.Ctx,
+		retryInterval:       cfg.RetryInterval,
+		tokenBufferInterval: cfg.TokenBufferInterval,
+		state:               cfg.State,
+		refresher:           cfg.Refresher,
+		clock:               cfg.Clock,
+		log:                 cfg.Log,
+	}
+
+	err := provider.init()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return provider, nil
+}
+
+func (r *RotatedAccessTokenProvider) init() error {
+	var err error
+
+	r.creds, err = r.state.GetCredentials(r.ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// GetAccessToken implements AccessTokenProvider()
+func (r *RotatedAccessTokenProvider) GetAccessToken() (string, error) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	return r.creds.AccessToken, nil
+}
+
+// RefreshLoop runs the credential refresh process.
+func (r *RotatedAccessTokenProvider) RefreshLoop() {
+	r.lock.RLock()
+	creds := r.creds
+	r.lock.RUnlock()
+
+	interval := r.getRefreshInterval(creds)
+
+	timer := r.clock.NewTimer(interval)
+	defer timer.Stop()
+	r.log.Infof("Will attempt token refresh in: %s", interval)
+
+	for {
+		select {
+		case <-r.ctx.Done():
+			r.log.Info("Shutting down")
+			return
+		case <-timer.Chan():
+			creds, _ := r.state.GetCredentials(r.ctx)
+
+			// Skip if the credentials are sufficiently fresh
+			// (in a HA setup another instance might have refreshed the credentials).
+			if creds != nil && !r.shouldRefresh(creds) {
+				r.lock.Lock()
+				r.creds = creds
+				r.lock.Unlock()
+
+				interval := r.getRefreshInterval(creds)
+				timer.Reset(interval)
+				r.log.Infof("Next refresh in: %s", interval)
+				continue
+			}
+
+			creds, err := r.refresh(r.ctx)
```

Comment:
It looks like there is a race condition where two HA instances can refresh the credentials. The logic above doesn't quite prevent this, as both instances could have called `GetCredentials` before either of them refresh the credentials. Will this cause the first instance that refreshed to have its credentials overwritten by the other instance refresh, or is this a race condition that we are ok with ignoring?

Reply:
A couple of possible issues with this simultaneous refresh/overwrite of credentials come to mind:
* Validity of refresh token - from my testing, these are not "one-time-use", you can use the same refresh token multiple times (in its validity period). This is probably by design: if one *used up* the refresh token, but failed to save the new credentials, the app would be broken for good and the user would need to re-initiate the authorization. :heavy_check_mark: 
* Validity of renewed credentials, i.e. whether refresh via the same "refresh token" invalidates previous credentials. Will need to test this one :question:

Reply:
Makes sense, can we add a comment so this doesn't get confused for a race condition down the road? Just looking at the code and comment above, I didn't realize the `shouldRefresh` logic is just an optimistic check, rather than core logic.